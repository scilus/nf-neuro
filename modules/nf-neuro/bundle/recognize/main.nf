process BUNDLE_RECOGNIZE {
    tag "$meta.id"
    label 'process_high'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(tractograms), path(transform), path(config), path(directory)

    output:
    tuple val(meta), path("*_cleaned.trk")            , emit: bundles
    tuple val(meta), path("*_bundles_mosaic_mqc.png")     , emit: mqc, optional: true
    tuple val(meta), path("*_bundles_stats_mqc.json")     , emit: global_mqc, optional: true
    path "versions.yml"                               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    // additional script arguments
    def run_qc = task.ext.run_qc ? task.ext.run_qc : false
    def minimal_vote_ratio = task.ext.minimal_vote_ratio ? "--minimal_vote_ratio " + task.ext.minimal_vote_ratio : ""
    def seed = task.ext.seed ? "--seed " + task.ext.seed : ""
    def rbx_processes = task.ext.rbx_processes ? "--processes " + task.ext.rbx_processes : "--processes 1"
    def outlier_alpha = task.ext.outlier_alpha ? "--alpha " + task.ext.outlier_alpha : ""


    """
    mkdir recobundles/
    scil_tractogram_segment_bundles.py ${tractograms} ${config} ${directory}/ ${transform} --inverse --out_dir recobundles/ \
        -v DEBUG $minimal_vote_ratio $seed $rbx_processes

    for bundle_file in recobundles/*.trk; do
        bname=\$(basename \${bundle_file} .trk)
        out_cleaned=${prefix}__\${bname}_cleaned.trk
        scil_bundle_reject_outliers.py \${bundle_file} "\${out_cleaned}" ${outlier_alpha}
    done

    if $run_qc;
    then
        # Take one file to generate temporary anat
        first_trk=(*_cleaned.trk)
        scil_tractogram_compute_density_map.py \${first_trk} tmp_anat.nii.gz --binary
        # Generate Mosaic for QC
        scil_viz_bundle_screenshot_mosaic.py tmp_anat.nii.gz *_cleaned.trk ${prefix}__bundles_mosaic_mqc.png --opacity_background 1
        rm -f tmp_anat.nii.gz

        # Generate JSON file stats
        for curr_bundle in *cleaned.trk; do
            bundle_name=\$(basename \${curr_bundle} _cleaned.trk)
            scil_bundle_shape_measures.py \${curr_bundle} --out_json \${bundle_name}__stats.json
        done
        scil_json_merge_entries.py *__stats.json ${prefix}__bundles_stats_mqc.json --keep_separate
        rm -f *__stats.json

    fi


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list --disable-pip-version-check --no-python-version-warning | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_tractogram_segment_bundles.py -h
    scil_bundle_reject_outliers.py -h

    # dummy output for single bundle
    touch ${prefix}__AF_L_cleaned.trk
    touch ${prefix}__bundles_mosaic_mqc.png
    touch ${prefix}__bundles_stats_mqc.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list --disable-pip-version-check --no-python-version-warning | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
