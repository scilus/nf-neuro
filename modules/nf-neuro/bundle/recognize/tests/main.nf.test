nextflow_process {

    name "Test Process BUNDLE_RECOGNIZE"
    script "../main.nf"
    process "BUNDLE_RECOGNIZE"

    tag "modules"
    tag "modules_nfcore"
    tag "bundle"
    tag "bundle/recognize"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "bundles.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("bundle - recognize - base") {

        config "./nextflow.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/bundle_all_1mm.trk"),
                        file("\${test_data_directory}/affine.txt"),
                        file("\${test_data_directory}/config.json"),
                        file("\${test_data_directory}/fibercup_atlas/")
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("bundle - recognize - params") {

        config "./nextflow_params.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/bundle_all_1mm.trk"),
                        file("\${test_data_directory}/affine.txt"),
                        file("\${test_data_directory}/config.json"),
                        file("\${test_data_directory}/fibercup_atlas/")
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.bundles,
                    file(process.out.mqc.get(0).get(1)).name,
                    file(process.out.global_mqc.get(0).get(1)).name,
                    process.out.versions,
                ).match() }
            )
        }

    }

}
