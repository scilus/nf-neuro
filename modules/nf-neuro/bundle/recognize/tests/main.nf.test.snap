{
    "bundle - recognize - params": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test__bundle_1_cleaned.trk:md5,4f5cb88d05f175d6b55509d0ad307ab3",
                        "test__bundle_2_cleaned.trk:md5,58b4c1381603d72ad6e72cc705b4606e",
                        "test__bundle_3_cleaned.trk:md5,8f24a8c534b6e2386bcc755bd8f0c4b9",
                        "test__bundle_4_cleaned.trk:md5,eafac0796051d6ada4bfc5a5c576c536",
                        "test__bundle_5_cleaned.trk:md5,0d512bcaaf49325415a1f4860a167f5e"
                    ]
                ]
            ],
            "test__bundles_mosaic_mqc.png",
            "test__bundles_stats_mqc.json",
            [
                "versions.yml:md5,083946242b74b84b09704b8be661eee2"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T20:25:45.388907231"
    },
    "bundle - recognize - base": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_1_cleaned.trk:md5,6a5c47dccc9f51087572a8de006183ae",
                            "test__bundle_2_cleaned.trk:md5,08e2bd3fe0432d4dd58b7db771c8080d",
                            "test__bundle_3_cleaned.trk:md5,313c4401e7771a966c26be3f73ea7730",
                            "test__bundle_4_cleaned.trk:md5,b2e8e5a6b9077d6a36552b6bfa579de4",
                            "test__bundle_5_cleaned.trk:md5,3b975ace291d6fd4fd434ce37cc2263a"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,083946242b74b84b09704b8be661eee2"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_1_cleaned.trk:md5,6a5c47dccc9f51087572a8de006183ae",
                            "test__bundle_2_cleaned.trk:md5,08e2bd3fe0432d4dd58b7db771c8080d",
                            "test__bundle_3_cleaned.trk:md5,313c4401e7771a966c26be3f73ea7730",
                            "test__bundle_4_cleaned.trk:md5,b2e8e5a6b9077d6a36552b6bfa579de4",
                            "test__bundle_5_cleaned.trk:md5,3b975ace291d6fd4fd434ce37cc2263a"
                        ]
                    ]
                ],
                "global_mqc": [
                    
                ],
                "mqc": [
                    
                ],
                "versions": [
                    "versions.yml:md5,083946242b74b84b09704b8be661eee2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-25T18:21:10.115548905"
    }
}